### PR TITLE
feat(acp): render running tool-call cards (pairs with SDK source dedup)

### LINKED ISSUE #1025: [ACP on Cloud] Refactor: decouple ACP tool-call streaming (live deltas) from persistence (one terminal event)
**Priority: P1** (perf/scale — do before real cloud load, not before bring-up). Part of #988.

## Problem
ACP re-emits the **full cumulative output** on every `ToolCallProgress` frame and **persists each one** as a distinct `ACPToolCallEvent` (`acp_agent.py:645-698`; `_default_callback` appends to the event log, `local_conversation.py:227-231`) → **O(n²)** on-disk storage + WebSocket relay per chatty tool call. On cloud the proxy relays *and* persists every cumulative frame across the network. The regular agent persists exactly one Action + one Observation per tool call.

## Not a streaming regression (verified against canvas `origin/main` + PRs)
The canvas **does not render these intermediate frames today**: `should-render-event.ts:88-90` only renders `completed`/`failed` ACP events — in-flight frames are *deliberately suppressed* (merged PR #142, port of OpenHands#14247, "so empty-args cards don't flash"). The card body is built from a single terminal event's `raw_input`/`raw_output` (`get-acp-tool-call-content.ts:99-131`); `handle-event-for-ui.ts:22-34` does a whole-object **replace**, not accumulation. So the intermediate cumulative frames are currently **persisted + relayed but never displayed** — pure O(n²) waste. Removing them regresses nothing.

Smooth **assistant-text** streaming is being added separately by canvas **#753** + SDK **#3376** via a **non-persisted `StreamingDeltaEvent`** (published straight to PubSub/WS, bypassing the persisted log; scoped to `content`/`reasoning_content` — no `tool_call_id`, i.e. not tool output). That's the precedent/pattern, not a conflict.

## The fix — decouple live from durable (this *enables* tool-output streaming, doesn't block it)
Model it on `on_token` / #753:
- [ ] **Live channel (non-persisted):** stream **incremental deltas** of tool output (the new chunk only — O(n) total) via an event published directly to `_pub_sub` so the WS relay forwards it to clients live (`event_service.py:714-733`, `sockets.py:434-442`). Either add a `tool_call_id` + output-delta field to `StreamingDeltaEvent` (coordinate with #753/#3376) or add a sibling `ACPToolCallProgressEvent`. **Send deltas, not the cumulative blob** — else the WS relay stays O(n²) even if disk drops to O(n).
- [ ] **Durable channel:** persist exactly **one** terminal `ACPToolCallEvent` per `tool_call_id` (the final merged state) — split today's single `on_event` path into live-progress + one terminal `on_event` (the SDK already maintains the cumulative `accumulated_tool_calls` dict, `acp_agent.py:631-666`).
- [ ] Delete the now-redundant downstream dedups this papered over: `RemoteConversation._add_event_unsafe` (`remote_conversation.py:347-389`), canvas `handle-event-for-ui.ts:22-34`, `resume_transcript._terminal_tool_indices` (`resume_transcript.py:211-244`), + the should-render gate (`should-render-event.ts:88-90`).
- [ ] Test against `resume_transcript` (relies on terminal-event semantics).

Net: one persisted event per tool call (maps 1:1 onto the regular action→observation model), O(n) live deltas, and a real path to **smooth tool-output streaming** instead of today's send-and-discard. Tradeoff (same as token streaming): a client reconnecting mid-tool-call replays only the terminal event from history, not the intermediate deltas. Implements pre-cloud refactor **P1-1**; coordinate with #753 / #3376.

### PR DESCRIPTION
**Pull request by simonrosenberg**
Closes #1025. Part of epic #988.

Attached a video to validate that the fix works
https://github.com/user-attachments/assets/1cc487cd-6a6f-4677-85b0-04bbcc1c130f

## Problem

ACP persisted one `ACPToolCallEvent` per `ToolCallProgress`, each carrying the full cumulative output → **O(n²)** storage + WebSocket relay per chatty tool call. The SDK side of this is fixed in **OpenHands/software-agent-sdk#3465**, which collapses each call to exactly two persisted events — an early **`started`** event and one **terminal** `completed`/`failed` event (the action→observation pair).

## Change

This is the canvas half: render the running card.

- **`should-render-event.ts`** — relax the ACP gate to render the card at **every** lifecycle stage (`return true`) instead of terminal-only. The `started` event now renders the card as **"running"** (no check mark, via `getACPToolCallResult`), and `handleEventForUI` replaces it in place by `tool_call_id` when the terminal event arrives — mirroring how an `ObservationEvent` supersedes its `ActionEvent`. The old terminal-only gate existed to hide the half-formed cards the SDK's per-progress fan-out flashed mid-stream; that fan-out is gone, so the running card is now a single clean event.
- **`handle-event-for-ui.ts`** — the `tool_call_id` merge is **kept** (now an O(1) `started`→`terminal` merge); comment refreshed to describe the action→observation semantics.
- **`acp-tool-call-event.ts`** — widen `ACPToolCallStatus` to include `pending` (the `started` event's status) and refresh the doc comment.

No new runtime dependency on the SDK PR is required to render correctly, but the UX (one clean running card → result, rather than a card that only appears on completion) is the point of pairing them.

## Tests

- `should-render-event.test.ts` updated: `in_progress` / `pending` / `null`-status events now render as running cards; `completed` / `failed` still render.
- Existing `handle-event-for-ui.test.ts` ACP merge tests and `event-message-acp-tool-call.test.tsx` (which already covers in-progress rendering) pass unchanged.
- Affected suites green (49 passed); `typecheck`, `eslint`, `prettier --check` clean.

## Pairs with

OpenHands/software-agent-sdk#3465 — dedup ACP tool-call progress at the source (O(n²)→O(1)).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
**🐳 Docker images for this PR**

• **GHCR package:** https://github.com/OpenHands/agent-canvas/pkgs/container/agent-canvas

| Component | Value |
|---|---|
| **Image** | `ghcr.io/openhands/agent-canvas` |
| **Architectures** | amd64, arm64 |
| **Agent Server** | `ghcr.io/openhands/agent-server:1.24.0-python` |
| **Automation** | `openhands-automation==1.0.0a5` |
| **Commit** | `144fd5fc78b0ecf4bc7b15049e0fecd977335685` |

**Pull (multi-arch manifest)**
```bash
# Multi-arch manifest — Docker automatically pulls the correct architecture
docker pull ghcr.io/openhands/agent-canvas:sha-144fd5f
```

**Run**
```bash
docker run -it --rm \
  -p 8000:8000 \
  ghcr.io/openhands/agent-canvas:sha-144fd5f
```

**All tags pushed for this build**
```
ghcr.io/openhands/agent-canvas:sha-144fd5f-amd64
ghcr.io/openhands/agent-canvas:acp-render-running-tool-cards-amd64
ghcr.io/openhands/agent-canvas:pr-1027-amd64
ghcr.io/openhands/agent-canvas:sha-144fd5f-arm64
ghcr.io/openhands/agent-canvas:acp-render-running-tool-cards-arm64
ghcr.io/openhands/agent-canvas:pr-1027-arm64
ghcr.io/openhands/agent-canvas:sha-144fd5f
ghcr.io/openhands/agent-canvas:acp-render-running-tool-cards
ghcr.io/openhands/agent-canvas:pr-1027
```

**About Multi-Architecture Support**
- Each tag (e.g., `sha-144fd5f`) is a **multi-arch manifest** supporting both **amd64** and **arm64**
- Docker automatically pulls the correct architecture for your platform
- Individual architecture tags (e.g., `sha-144fd5f-amd64`) are also available if needed

### CODE REVIEW

File: src/types/agent-server/core/events/acp-tool-call-event.ts [hunk 1]
```diff
@@ -8,11 +8,17 @@ import { BaseEvent } from "../base/event";
...
- * will emit multiple events for the same ``tool_call_id`` as the status
- * transitions from ``in_progress`` to a terminal value.
+ * Status of an ACP tool call. The SDK persists two events per
+ * ``tool_call_id``: an early ``started`` event (``pending`` / ``in_progress``)
+ * and one terminal event (``completed`` / ``failed``) — the
+ * action->observation pair. Non-terminal statuses render the card as
+ * "running"; see ``getACPToolCallResult``.
  */
-export type ACPToolCallStatus = "in_progress" | "completed" | "failed";
+export type ACPToolCallStatus =
```
[the remark this answers is not in the document]
Comment (simonrosenberg) [reply src/types/agent-server/core/events/acp-tool-call-event.ts:17]: Done in 3d009eb1 — updated the `getACPToolCallResult` JSDoc to say a non-terminal call (`pending` / `in_progress`) falls through to `undefined` (running card).

File: src/utils/handle-event-for-ui.ts [hunk 2]
```diff
@@ -9,9 +9,11 @@ import {
  * Replaces actions with observations when they arrive (so UI shows observation instead of action)
  * Exception: ThinkAction is NOT replaced because the thought content is in the action, not in the observation
  *
- * ACPToolCallEvent dedup: multiple events share a ``tool_call_id`` as an ACP
- * tool call progresses (in_progress → completed / failed). Collapse them to
- * the latest state at the original position so the card updates in place.
+ * ACPToolCallEvent merge: the SDK emits two events per ``tool_call_id`` — an
+ * early ``started`` event (in_progress) and one terminal (completed / failed)
```
[the remark this answers is not in the document]
Comment (simonrosenberg) [reply src/utils/handle-event-for-ui.ts]: Fixed in 9c6e060a — the comment now reads `(``pending`` / ``in_progress``)`. Also addressed the test-gap note: added a dedicated `getACPToolCallResult` unit suite (pending/in_progress/null → undefined, failed/is_error → error, completed → success).

File: src/utils/handle-event-for-ui.ts [hunk 3]
```diff
@@ -9,9 +9,12 @@ import {
  * Replaces actions with observations when they arrive (so UI shows observation instead of action)
  * Exception: ThinkAction is NOT replaced because the thought content is in the action, not in the observation
  *
- * ACPToolCallEvent dedup: multiple events share a ``tool_call_id`` as an ACP
- * tool call progresses (in_progress → completed / failed). Collapse them to
- * the latest state at the original position so the card updates in place.
+ * ACPToolCallEvent merge: the SDK emits two events per ``tool_call_id`` — an
+ * early ``started`` event (``pending`` / ``in_progress``) and one terminal
+ * (completed / failed) event, the action->observation pair for a tool call.
+ * Replace the started
```
[the remark this answers is not in the document]
Comment (simonrosenberg) [reply src/utils/handle-event-for-ui.ts]: Done in 5640fee8 — reflowed onto continuous lines.

**Review by enyst — APPROVED** [review]
Thank you!

The PR has only ACP behavior slightly modified, that is, the fix itself. So it doesn’t pose regression risks.
